streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java

```diff
@@ -0,0 +1,303 @@
+    private KTable<String, Long> table;
+    final Map<String, String> results = new HashMap<>();
+    private ForeachAction<String, String> foreachAction;
+
+    @Before
+    public void before() {
+        testNo++;
+        builder = new KStreamBuilder();
+        createTopics();
+        streamsConfiguration = new Properties();
+        final String applicationId = "globalOne-table-test-" + testNo;
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, applicationId);
+        streamsConfiguration
+                .put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.ZOOKEEPER_CONNECT_CONFIG, CLUSTER.zKConnectString());
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        globalTable = builder.globalTable(Serdes.Long(), Serdes.String(), globalOne, globalStore);
+        stream = builder.stream(Serdes.String(), Serdes.Long(), inputStream);
+        table = builder.table(Serdes.String(), Serdes.Long(), inputTable, "table");
+        foreachAction = new ForeachAction<String, String>() {
+            @Override
+            public void apply(final String key, final String value) {
+                results.put(key, value);
+            }
+        };
+    }
+
+    @After
+    public void whenShuttingDown() throws IOException {
+        if (kafkaStreams != null) {
+            kafkaStreams.close();
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void shouldKStreamGlobalKTableLeftJoin() throws Exception {
+        final KStream<String, String> streamTableJoin = stream.leftJoin(globalTable, keyMapper, joiner);
+        streamTableJoin.foreach(foreachAction);
+        runAndVerifyLeftJoin(inputStream);
+    }
+
+    @Test
+    public void shouldKStreamGlobalKTableJoin() throws Exception {
+        final KStream<String, String> streamTableJoin = stream.join(globalTable, keyMapper, joiner);
+        streamTableJoin.foreach(foreachAction);
+        runAndVerifyJoin(inputStream);
+    }
+
+    @Test
+    public void shouldKTableGlobalKTableLeftJoin() throws Exception {
+        final KTable<String, String> tableGlobalTableJoin = table.leftJoin(globalTable, keyMapper, joiner);
+        tableGlobalTableJoin.foreach(foreachAction);
+        runAndVerifyLeftJoin(inputTable);
+    }
+
+    @Test
+    public void shouldKTableGlobalKTableJoin() throws Exception {
+        final KTable<String, String> tableGlobalTableJoin = table.join(globalTable, keyMapper, joiner);
+        tableGlobalTableJoin.foreach(foreachAction);
+        runAndVerifyJoin(inputTable);
+    }
+
+    private void runAndVerifyJoin(final String topic) throws InterruptedException, ExecutionException {
```

Comment:
That is exactly what I had in mind with my previous comment... I just thought we might be able to not send the old value at all -- but you example shows that we need to send it.

One way to fix it, is to remember the old value on the triggering KTable side (I guess we need another store for this...). Thus, instead of looking up old value in GlobalKTable, we look it up the the new store. Not sure if there is a better way to do it.